File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Fp8Quantize.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.Getter;
import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.List;

/**
 * Quantize a float tensor to FP8 E4M3 representation with scale factors.
 * <p>
 * Supports three quantization modes:
 * <ul>
 *   <li>Per-tensor (mode=0): single scale for the entire tensor</li>
 *   <li>Per-token (mode=1): one scale per row (default)</li>
 *   <li>Per-group (mode=2): one scale per group of {@code groupSize} elements</li>
 * </ul>
 * <p>
 * Inputs:
 * <ul>
 *   <li>0: input [numTokens, hiddenDim] float/half/bf16</li>
 * </ul>
 * <p>
 * Outputs:
 * <ul>
 *   <li>0: quantized [numTokens, hiddenDim] INT8 (FP8 E4M3 values)</li>
 *   <li>1: scales - [numTokens] for per-token, [numTokens, numGroups] for per-group</li>
 * </ul>
 */
@NoArgsConstructor
public class Fp8Quantize extends DynamicCustomOp {

    public static final int MODE_PER_TENSOR = 0;
    public static final int MODE_PER_TOKEN = 1;
    public static final int MODE_PER_GROUP = 2;

    public static final int FORMAT_E4M3 = 0;
    public static final int FORMAT_E5M2 = 1;

    @Getter private int mode = MODE_PER_TOKEN;
    @Getter private int groupSize = 128;
    @Getter private int fp8Format = FORMAT_E4M3;

    public Fp8Quantize(INDArray input) {
        this(input, MODE_PER_TOKEN, 128, FORMAT_E4M3);
    }

    public Fp8Quantize(INDArray input, int mode, int groupSize, int fp8Format) {
        super(new INDArray[]{input}, null);
        this.mode = mode;
        this.groupSize = groupSize;
        this.fp8Format = fp8Format;
        addIArgument((long) mode, (long) groupSize, (long) fp8Format);
    }

    public Fp8Quantize(SameDiff sameDiff, SDVariable input, int mode, int groupSize, int fp8Format) {
        super(null, sameDiff, new SDVariable[]{input}, false);
        this.mode = mode;
        this.groupSize = groupSize;
        this.fp8Format = fp8Format;
        addIArgument((long) mode, (long) groupSize, (long) fp8Format);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (iArguments.size() > 0) this.mode = iArguments.get(0).intValue();
        if (iArguments.size() > 1) this.groupSize = iArguments.get(1).intValue();
        if (iArguments.size() > 2) this.fp8Format = iArguments.get(2).intValue();
    }

    @Override
    public String opName() {
        return "fp8_quantize";
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        return Arrays.asList(DataType.INT8, DataType.FLOAT);
    }

    @Override
    public int getNumOutputs() {
        return 2;
    }
}

File: libnd4j/include/ops/declarable/generic/nn/fp8_quantize.cpp
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_fp8_quantize)

#include <ops/declarable/headers/nn.h>
#include <ops/declarable/helpers/per_token_quant.h>

namespace sd {
namespace ops {

/**
 * fp8_quantize - Quantize float tensor to FP8 E4M3 representation.
 *
 * Supports two modes:
 *   mode=0: Per-tensor quantization (single scale for entire tensor)
 *   mode=1: Per-token (per-row) quantization (one scale per row)
 *   mode=2: Per-group quantization (one scale per group of groupSize elements)
 *
 * Input:
 *   0: input tensor [numTokens, hiddenDim] - float/half/bf16
 *
 * Output:
 *   0: quantized tensor [numTokens, hiddenDim] - INT8 (stores FP8 E4M3 values)
 *   1: scale factors - [1] for per-tensor, [numTokens] for per-token,
 *                       [numTokens, numGroups] for per-group
 *
 * Integer args:
 *   0: mode (0=per-tensor, 1=per-token, 2=per-group, default: 1)
 *   1: groupSize (for mode=2, default: 128)
 *   2: fp8Format (0=E4M3, 1=E5M2, default: 0)
 */
CUSTOM_OP_IMPL(fp8_quantize, 1, 2, false, 0, 0) {
    auto input = INPUT_VARIABLE(0);
    auto quantized = OUTPUT_VARIABLE(0);
    auto scales = OUTPUT_VARIABLE(1);

    int mode = block.getIArguments()->size() > 0 ? INT_ARG(0) : 1;
    int groupSize = block.getIArguments()->size() > 1 ? INT_ARG(1) : 128;

    REQUIRE_TRUE(input->rankOf() == 2, 0,
                 "fp8_quantize: input must be rank 2 [numTokens, hiddenDim], got %lld",
                 (long long)input->rankOf());

    if (mode == 2) {
        // Per-group
        helpers::perTokenGroupQuant(block.launchContext(), input, quantized, scales, groupSize, true);
    } else {
        // Per-token (mode 0 and 1 both use perTokenQuantFp8;
        // for mode=0/per-tensor, the caller reshapes to [1, totalElements])
        helpers::perTokenQuantFp8(block.launchContext(), input, quantized, scales);
    }

    return sd::Status::OK;
}

DECLARE_TYPES(fp8_quantize) {
  getOpDescriptor()->addTraits(OP_TRAIT_UNARY_ELEMENTWISE | OP_TRAIT_FULLY_WRITING);
    getOpDescriptor()->setAllowedInputTypes({ALL_FLOATS});
    getOpDescriptor()->setAllowedOutputTypes(0, {INT8});
    getOpDescriptor()->setAllowedOutputTypes(1, {FLOAT32});
}

DECLARE_SHAPE_FN(fp8_quantize) {
    auto inShape = inputShape->at(0);
    auto numTokens = shape::sizeAt(inShape, static_cast<LongType>(0));
    auto hiddenDim = shape::sizeAt(inShape, static_cast<LongType>(1));

    int mode = block.getIArguments()->size() > 0 ? INT_ARG(0) : 1;
    int groupSize = block.getIArguments()->size() > 1 ? INT_ARG(1) : 128;

    // Output 0: quantized [numTokens, hiddenDim] INT8
    LongType quantShape[2] = {numTokens, hiddenDim};
    auto quantizedShape = ConstantShapeHelper::getInstance().createShapeInfo(
        INT8, 'c', 2, quantShape);

    // Output 1: scales
    LongType* scalesShape;
    if (mode == 2) {
        // Per-group: [numTokens, numGroups]
        LongType numGroups = (hiddenDim + groupSize - 1) / groupSize;
        LongType groupShape[2] = {numTokens, numGroups};
        scalesShape = ConstantShapeHelper::getInstance().createShapeInfo(
            FLOAT32, 'c', 2, groupShape);
    } else {
        // Per-token: [numTokens]
        scalesShape = ConstantShapeHelper::getInstance().vectorShapeInfo(numTokens, FLOAT32);
    }

    return SHAPELIST(quantizedShape, scalesShape);
}

}  // namespace ops
}  // namespace sd

#endif
